### PR TITLE
[receiver/awss3receiver] Initial implementation

### PR DESCRIPTION
**Pull request by adcharre**
**Description:** This is the initial implementation of the AWS S3 receiver. The receiver can load trace from an S3 bucket starting at the configured time until the stop time. Json and protobuf formats are supported along with gzip compression.

**Link to tracking Issue:** #30750

**Testing:** Unit tests added and read real trace from an S3 bucket.

**Documentation:** None added

### CODE REVIEW

File: receiver/awss3receiver/s3reader.go
Comment (crobert-1) [inline receiver/awss3receiver/s3reader.go:119]: If golang's `net/http` package, the client is required to close the body of get request responses to avoid leaking goroutines. Is that true for AWS's s3 library here?

File: receiver/awss3receiver/s3intf.go
Comment (crobert-1) [inline receiver/awss3receiver/s3intf.go:44]: Should the AWS partition be user-configurable? Will "aws" always work for configuration, or at least work for the large majority of users?

**Review by atoulme — APPROVED** [review]

File: receiver/awss3receiver/receiver.go
Comment (crobert-1) [inline receiver/awss3receiver/receiver.go:74]: Should an error be logged in the case of an invalid or unexpected key string format? I'm wondering about the case of users wondering why expected data isn't coming through because of an unsupported format.

File: receiver/awss3receiver/s3reader.go
Comment (crobert-1) [inline receiver/awss3receiver/s3reader.go:45]: Out of ignorance, why is this time granularity referred to as `S3 partition`? I don't have much context here, so I don't understand the relevance to partition.

**Review by crobert-1 — COMMENTED** [review]
I didn't get a chance to look through tests yet, just an initial pass over functionality. I don't have much AWS experience, so some questions may be pretty elementary. 👍

File: receiver/awss3receiver/receiver.go
Comment (adcharre) [reply receiver/awss3receiver/receiver.go:72]: Agreed - at this stage just want to get the basics in and then add the ability to use extensions to unmarshall data next.

File: receiver/awss3receiver/s3reader.go
Comment (adcharre) [reply receiver/awss3receiver/s3reader.go:45]: The reason for S3Partition name is due to this reflecting what is already being used by the S3 exporter. It makes sense to use the same config names when the receiver is the complement of the exporter.

File: receiver/awss3receiver/s3intf.go
Comment (adcharre) [reply receiver/awss3receiver/s3intf.go:44]: This one is a bit of a magic value taken from localstacks configuration of the aws sdk go v2: https://docs.localstack.cloud/user-guide/integrations/sdks/go/

Comment (crobert-1) [reply receiver/awss3receiver/s3intf.go:44]: Do you think there would ever be value in allowing the user to configure this? I'm not very familiar with AWS and how it's using partition ID here.

It doesn't have to be done in this PR if it is valuable, just want to get a better idea of what's going on here.

Comment (adcharre) [reply receiver/awss3receiver/s3intf.go:44]: It's always possible - after all why make it configurable at the AWS SDK level if it's only ever going to equal "aws"!
After a bit more digging on this I found the partition IDs listed in the v1 sdk: https://github.com/aws/aws-sdk-go/blob/8888dae11935b652781010304c637406a168475a/aws/endpoints/defaults.go#L11
It won't take much to add the configuration option - I'll do it now.

**Review by crobert-1 — COMMENTED** [review]
_I see I've got a couple open comments still, so I'll wait to review tests until after those are responded to 👍_

File: receiver/awss3receiver/s3reader.go
Comment (adcharre) [reply receiver/awss3receiver/s3reader.go:18]: These are interfaces so are references - the reason for using these is to allow mocking of the AWS SDK for unit testing.

File: receiver/awss3receiver/s3reader.go
Comment (adcharre) [reply receiver/awss3receiver/s3reader.go]: I've moved this to the readAll function where timeStep should make more sense. It's defined above and used by the for loop which should make it clear.

File: receiver/awss3receiver/README.md
Comment (adcharre) [reply receiver/awss3receiver/README.md:30]: Bah - it's one of those things where I did it in Goland and it look correct but it's being clever and so doesn't show you the true layout. I wonder if there's a reformat option for Markdown, hmmm...

Comment (adcharre) [reply receiver/awss3receiver/README.md]: Turns out there is a reformat table option!

**Comment by adcharre** [comment]
@crobert-1 
> _I see I've got a couple open comments still, so I'll wait to review tests until after those are responded to 👍_

I think I've been through all your original comments and resolved them now.

File: receiver/awss3receiver/s3intf.go
Comment (crobert-1) [reply receiver/awss3receiver/s3intf.go:51]: I see this was marked as resolved, but I don't think it was changed. Are you planning on making this update, or is there a reason not to?

Comment (adcharre) [reply receiver/awss3receiver/s3intf.go]: My mistake - updated now.

**Review by crobert-1 — APPROVED** [review]
Looks good to me, thanks for your patience with all of my questions and comments!

**Comment by atoulme** [comment]
Just need a conflict fix, sorry.

**Comment by adcharre** [comment]
> Just need a conflict fix, sorry.

Updated (again ...)

**Comment by atoulme** [comment]
Ready to go!

**Comment by dmitryax** [comment]
Hey @adcharre, sorry for the delay with merging. Can you please rebase again?

**Comment by adcharre** [comment]
> Hey @adcharre, sorry for the delay with merging. Can you please rebase again?

@dmitryax done!

**Comment by adcharre** [comment]
Rebased again ... @dmitryax any chance this can go in soon while I just love resolving merge conflicts I would like to see this merged :-)

**Comment by xThomo** [comment]
@atoulme / @dmitryax is there a reason holding up the merge of this PR? I see others being merged over the past couple of weeks.

File: receiver/awss3receiver/receiver.go
Comment (crobert-1) [inline receiver/awss3receiver/receiver.go:40]: Sorry, one more comment, I just learned from the [component specification](https://pkg.go.dev/go.opentelemetry.io/collector/component#Component) that background tasks are supposed to be run using a new context from `context.Background`, not the passed in context. (See comment on `Start`)

Comment (adcharre) [reply receiver/awss3receiver/receiver.go]: Updated.

File: receiver/awss3receiver/receiver.go
Comment (crobert-1) [inline receiver/awss3receiver/receiver.go:39]:
```suggestion
func (r *awss3TraceReceiver) Start(_ context.Context, _ component.Host) error {
```
I believe lint may fail on this line if `ctx` isn't used.

**Comment by atoulme** [comment]
> @atoulme / @dmitryax is there a reason holding up the merge of this PR? I see others being merged over the past couple of weeks.

No reason that I can think of. I have approved this PR and marked it ready to merge. I cannot merge PRs, maintainers can. They are busy and this is a big PR.

File: receiver/awss3receiver/receiver.go
Comment (adcharre) [reply receiver/awss3receiver/receiver.go:39]: It won't but for the not very nice reason that I've reused the `ctx` variable on the next line. I'll push another small update to make it clearer that the context passed in is not used.

**Review by crobert-1 — APPROVED** [review]

File: receiver/awss3receiver/receiver.go
Comment (crobert-1) [reply receiver/awss3receiver/receiver.go]: Thank you!

File: receiver/awss3receiver/go.mod
Comment (atoulme) [inline receiver/awss3receiver/go.mod:62]:
```suggestion
	go.opentelemetry.io/collector/config/configtelemetry v0.99.1-0.20240502202854-2875844e3c35 // indirect
```

**Comment by atoulme** [comment]
@adcharre I tried to help but I don't think I got it right, mind taking a look?

**Comment by adcharre** [comment]
@dmitryax rebased ready for merging

**Pull request merged**
